File: python_bot/bot/key_manager.py
import os
import json
import base64
from Crypto.Cipher import AES
import os
from dotenv import load_dotenv

env_path = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))), ".env.local")
load_dotenv(env_path)

decrypted_keys = None
current_index = 0

def decrypt(encrypted_str: str, secret_hex: str) -> str:
    parts = encrypted_str.split(':')
    if len(parts) != 3:
        raise ValueError("Invalid encrypted string format. Expected iv:authTag:ciphertext")
        
    iv_hex, auth_tag_hex, ciphertext_hex = parts

    key = bytes.fromhex(secret_hex)
    iv = bytes.fromhex(iv_hex)
    auth_tag = bytes.fromhex(auth_tag_hex)
    ciphertext = bytes.fromhex(ciphertext_hex)

    cipher = AES.new(key, AES.MODE_GCM, nonce=iv)
    decrypted = cipher.decrypt_and_verify(ciphertext, auth_tag)
    
    return decrypted.decode('utf-8')

def load_keys() -> list[str]:
    global decrypted_keys
    if decrypted_keys is not None:
        return decrypted_keys

    # Read keys from environment
    env_keys = os.getenv('GEMINI_API_KEYS')
    if env_keys:
        decrypted_keys = [k.strip() for k in env_keys.split(',') if k.strip()]
        print(f"[KeyManager] [SUCCESS] Successfully loaded {len(decrypted_keys)} API keys for rotation.")
        return decrypted_keys
    
    raise ValueError("GEMINI_API_KEYS environment variable is missing or empty. Please set it securely in your .env.local file.")

def get_next_api_key() -> str:
    global current_index
    keys = load_keys()
    key = keys[current_index]
    current_index = (current_index + 1) % len(keys)
    return key

def get_key_count() -> int:
    return len(load_keys())

def reset_rotation() -> None:
    global current_index
    current_index = 0

File: src/lib/bot/key-manager.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unused-vars */
/**
 * Gemini API Key Manager
 * 
 * - Decrypts AES-256-GCM encrypted keys from environment variables (server-side only)
 * - Provides round-robin rotation across all 8 keys per request
 * - Keys are NEVER exposed to the browser (no NEXT_PUBLIC_ prefix)
 */

import { createDecipheriv } from 'crypto';

// ─── Internal State ───
let decryptedKeys: string[] | null = null;
let currentIndex = 0;

/**
 * Decrypt a single AES-256-GCM encrypted string.
 * Expected format: iv:authTag:ciphertext (all hex-encoded)
 */
function decrypt(encryptedStr: string, secretHex: string): string {
  const [ivHex, authTagHex, ciphertext] = encryptedStr.split(':');

  const key = Buffer.from(secretHex, 'hex');
  const iv = Buffer.from(ivHex, 'hex');
  const authTag = Buffer.from(authTagHex, 'hex');

  const decipher = createDecipheriv('aes-256-gcm', key, iv);
  decipher.setAuthTag(authTag);

  let decrypted = decipher.update(ciphertext, 'hex', 'utf8');
  decrypted += decipher.final('utf8');

  return decrypted;
}

/**
 * Lazily loads and decrypts all API keys on first call.
 * Throws if env vars are missing or decryption fails.
 */
function loadKeys(): string[] {
  if (decryptedKeys) return decryptedKeys;

  const secret = process.env.GEMINI_ENCRYPTION_SECRET;
  const encryptedJson = process.env.GEMINI_ENCRYPTED_KEYS;

  if (!secret || !encryptedJson) {
    throw new Error(
      '[KeyManager] Missing GEMINI_ENCRYPTION_SECRET or GEMINI_ENCRYPTED_KEYS in environment variables.'
    );
  }

  try {
    const encryptedArray: string[] = JSON.parse(encryptedJson);
    decryptedKeys = encryptedArray.map((enc) => decrypt(enc, secret));
    console.log(`[KeyManager] ✅ Successfully loaded ${decryptedKeys.length} API keys.`);
    return decryptedKeys;
  } catch (error: any) {
    throw new Error(`[KeyManager] Failed to decrypt API keys: ${error.message}`);
  }
}

/**
 * Returns the next API key using round-robin rotation.
 * Each call returns a different key to distribute load and avoid rate limits.
 */
export function getNextApiKey(): string {
  const keys = loadKeys();
  const key = keys[currentIndex];
  currentIndex = (currentIndex + 1) % keys.length;
  return key;
}

/**
 * Returns the total number of available API keys.
 */
export function getKeyCount(): number {
  return loadKeys().length;
}

/**
 * Resets the rotation index (useful for testing).
 */
export function resetRotation(): void {
  currentIndex = 0;
}
